provisioner/windows-restart/provisioner.go

```diff
@@ -174,29 +174,56 @@ WaitLoop:
 }
 
 var waitForCommunicator = func(p *Provisioner) error {
+	runCustomRestartCheck := true
+	if p.config.RestartCheckCommand == DefaultRestartCheckCommand {
+		runCustomRestartCheck = false
+	}
+	// This command is configurable by the user to make sure that the
+	// vm has met their necessary criteria for having restarted. If the
+	// user doesn't set a special restart command, we just run the
+	// default as cmdModuleLoad below.
+	cmdRestartCheck := &packer.RemoteCmd{Command: p.config.RestartCheckCommand}
+	log.Printf("Checking that communicator is connected with: '%s'",
+		cmdRestartCheck.Command)
 	for {
-		cmd := &packer.RemoteCmd{Command: p.config.RestartCheckCommand}
-		var buf, buf2 bytes.Buffer
-		cmd.Stdout = &buf
-		cmd.Stdout = io.MultiWriter(cmd.Stdout, &buf2)
 		select {
 		case <-p.cancel:
 			log.Println("Communicator wait canceled, exiting loop")
 			return fmt.Errorf("Communicator wait canceled")
 		case <-time.After(retryableSleep):
 		}
-
-		log.Printf("Checking that communicator is connected with: '%s'", cmd.Command)
-
-		err := cmd.StartWithUi(p.comm, p.ui)
-
-		if err != nil {
-			log.Printf("Communication connection err: %s", err)
-			continue
+		if runCustomRestartCheck == true {
+			// run user-configured restart check
+			err := cmdRestartCheck.StartWithUi(p.comm, p.ui)
+			if err != nil {
+				log.Printf("Communication connection err: %s", err)
+				continue
+			}
+			log.Printf("Connected to machine")
+			runCustomRestartCheck = false
 		}
 
-		log.Printf("Connected to machine")
-		stdoutToRead := buf2.String()
+		// This is the non-user-configurable check that powershell
+		// modules have loaded.
+
+		// If we catch the restart in just the right place, we will be able
+		// to run the restart check but the output will be an error message
+		// about how it needs powershell modules to load, and we will start
+		// provisioning before powershell is actually ready.
+		// In this next check, we parse stdout to make sure that the command is
+		// actually running as expected.
+		var buf bytes.Buffer
+		cmdModuleLoad := &packer.RemoteCmd{
+			Command: DefaultRestartCheckCommand,
+			Stdin:   nil,
+			Stdout:  &buf,
+			Stderr:  &buf}
```

Comment:
I don't believe it's safe to use `buf` twice, here. I also don't think we need to worry about stderr

Reply:
We follow this pattern elsewhere in the code, and if I don't throw something into stderr we log a warning about how we're not reading from stderr.  Not a big deal, but could be a red herring for someone trying to troubleshoot a problem.

Reply:
hmm, that's concerning. The ssh client we use has a writer called `singleWriter` that uses a mutex during write to make sure it's concurrently write-safe.

writes to stderr and stdout are done concurrently, and bytes.buffer is certainly not thread safe